### ISSUE TITLE
[C# Feature] Convenient combination of foreach and LINQ

### ISSUE THREAD

**Issue by dsaf** [dotnet/roslyn#1938]
I would argue that _embedded_ is a better term for how the syntax feels at the moment.

Currently:

``` csharp
foreach (var item in from xItem in new[] {"One", "Two", "Three" } 
                     where xItem == "Two"
                     select xItem)
{
    Debug.WriteLine(item);
}
```

Suggested:

``` csharp
foreach (from item in new[] {"One", "Two", "Three" } 
         where item == "Two"
         select item)
{
    Debug.WriteLine(item.ToString());
}
```

**Comment by HaloFour** [comment]
How about:

```cs
from item in new[] {"One", "Two", "Three" } 
where item == "Two"
foreach {
   Debug.WriteLine(item.ToString());
}
```

**Comment by bondsbw** [comment]
Thank you @dsaf for this suggestion.  And thank you @HaloFour, I love that syntax!

**Comment by apskim** [comment]
OK, if we are opening the Pandora box of syntactic sugar now, let's not stop at the frosting level:
```cs
foreach item in new[] {"One", "Two", "Three" } 
  where item == "Two"
{
   Debug.WriteLine(item.ToString());
}
```

**Comment by dsaf** [comment]
Both options look like extending LINQ rather than integrating it. To me having two ```foreach```'s - with and without round brackets is weird. But to be fair your suggestions are quite close to what X++ is already using and it seems to be working: https://msdn.microsoft.com/en-us/library/aa558063.aspx

**Comment by bondsbw** [comment]
Actually @dsaf, I'd say that @apskim's approach is merely your original proposal, coupled with two syntax changes.

The first change is to remove the ```select item``` portion.  I feel this is necessary anyway, because the name ```item``` is already in scope from the ```from``` clause and the ```select``` portion does nothing.

The second change removes the parentheses.  This could be a change to the general case of ```foreach```, allowing

```c#
foreach (var x in myCollection)
{
   // ...
}
```
to be rewritten as
```c#
foreach var x in myCollection
{
   // ...
}
```
I think removing the parentheses would require a braced body, otherwise perhaps the syntax would become ambiguous.

**Comment by bondsbw** [comment]
Actually I missed that the ```from``` is removed in the shortened syntax.  For that matter... why do we need the ```var``` in the ```foreach``` clause?  That could be implicit.

```c#
foreach x in myCollection
{
    //...
}
```

**Comment by apskim** [comment]
Just to clarify: my proposal was mostly sarcastic :)
I personally think that such syntactic sugar for the sake of sugaring merely complicates the language and breaks the consistent interplay of various language constructs.

**Comment by bondsbw** [comment]
I don't think it's just for "the sake of sugaring".  I feel there are clear advantage to reducing:

```c#
var filteredCollection =
    from t in myCollection
    where t < 7
    select t;

foreach (var x in filteredCollection)
{
    //...
}
```
to
```c#
foreach x in myCollection
where x < 7
{
    //...
}
```

Besides it obviously reducing the amount of code, it reduces the number of variables by removing ```t``` and ```filteredCollection```.

Also imagine if it were more complicated (example based on [MSDN LINQ samples](https://msdn.microsoft.com/en-us/library/gg509064.aspx)):

```c#
var query = 
    from c in svcContext.ContactSet
    join a in svcContext.AccountSet
        on c.ContactId equals a.PrimaryContactId.Id
    where c.LastName == "Wilcox" || c.LastName == "Andrews"
    where a.Address1_Telephone1.Contains("(206)")
        || a.Address1_Telephone1.Contains("(425)")
    select new
    {
        Contact = new Contact
        {
            FirstName = c.FirstName,
            LastName = c.LastName,
        },
        Account = new Account
        {
            Address1_Telephone1 = a.Address1_Telephone1
        }
    };
foreach (var record in query)
{
    Console.WriteLine("Contact Name: {0} {1}",
        record.Contact.FirstName, record.Contact.LastName);
    Console.WriteLine("Account Phone: {0}",
        record.Account.Address1_Telephone1);
}
```
could become more succinctly
```c#
foreach c in svcContext.ContactSet
    join a in svcContext.AccountSet
        on c.ContactId equals a.PrimaryContactId.Id
    where c.LastName == "Wilcox" || c.LastName == "Andrews"
    where a.Address1_Telephone1.Contains("(206)")
        || a.Address1_Telephone1.Contains("(425)")
{
    Console.WriteLine("Contact Name: {0} {1}", c.FirstName, c.LastName);
    Console.WriteLine("Account Phone: {0}", a.Account.Address1_Telephone1);
}
```
because it helps to eliminate some needs for anonymous types.

**Comment by aluanhaddad** [comment]
I don't think this adds much. 
Also the code:
```CSharp
var query = 
    from c in svcContext.ContactSet
    join a in svcContext.AccountSet
        on c.ContactId equals a.PrimaryContactId.Id
    where c.LastName == "Wilcox" || c.LastName == "Andrews"
    where a.Address1_Telephone1.Contains("(206)")
        || a.Address1_Telephone1.Contains("(425)")
    select new
    {
        Contact = new Contact
        {
            FirstName = c.FirstName,
            LastName = c.LastName,
        },
        Account = new Account
        {
            Address1_Telephone1 = a.Address1_Telephone1
        }
    };
```
could already be shortened to:
```CSharp
var query = 
    from c in svcContext.ContactSet
    join a in svcContext.AccountSet
        on c.ContactId equals a.PrimaryContactId.Id
    where c.LastName == "Wilcox" || c.LastName == "Andrews"
    where a.Address1_Telephone1.Contains("(206)")
        || a.Address1_Telephone1.Contains("(425)")
    select new
    {
        c.FirstName,
        c.LastName,
        a.Address1_Telephone1
    };
```
Other proposals for expanding LINQ syntax, such as the one proposing the ability to specify an IEqualityComparer inline (#100) add real value.

**Comment by bondsbw** [comment]
@aluanhaddad 

> Since c is the range variable following ```foreach```, I would not expect a to be in scope. It would only make sense if foreach was the last clause.

Keeping range variables in scope is, in my mind, a natural part of the proposal.  I don't really see the confusion, since range variables are similarly in scope for any ```select``` including the ```select (Action)...``` workaround you mentioned.

**Comment by HaloFour** [comment]
The problem with expressing the `foreach` clause at the beginning of the query is that the range variables can go in and out of scope.  For example:

```cs
var q = from x in Enumerable.Range(0, 10)
        select new String('X', x) into letters
        from y in letters
        select x; // compiler error, x is not in scope
```

Also:

```cs
var q = from x in Enumerable.Range(0, 10)
        select new String('X', x) into letters // here x is an int
        from x in letters
        select x; // here x is a char
```

This is why I would prefer that terminating clauses like `foreach` occur at the end of the query.

**Comment by aluanhaddad** [comment]
@HaloFour The confusion I was referring to was precisely due to the ```foreach``` clause being at the beginning. 
As you eloquently stated:
> The problem with expressing the foreach clause at the beginning of the query is that the range variables can go in and out of scope.

**Comment by alrz** [comment]
Basically, what @HaloFour have suggested, but with a `do` and an *embedded-statement* following.
```cs
from item in new[] {"One", "Two", "Three" } 
where item == "Two"
do Debug.WriteLine(item.ToString());
```
Then *query-expression* should be part of *embedded-statement* I guess, and in that context it must ends with a *do-clause* instead of *select-* or *group-clause*.

**Comment by aluanhaddad** [comment]
@alrz Under peril of being accused of bikeshedding, I want to save that ```foreach``` would be a far superior to ```do```.

**Comment by gafter** [comment]
So are you folks all fine with *not* being able to return/break from a loop?

**Comment by HaloFour** [comment]
@gafter Why wouldn't that be possible?  If like @alrz suggested this keyword was followed by an *embedded-statement* it would seem reasonable that said statement, like any `for`/`foreach`/`while`/`do` loop would be capable of using `continue`, `break` or `return`.

```cs
from number in Enumerable.Range(0, 10)
foreach {
    if (number < 5) continue;
    else if (number == 8) break;
    else if (number == 9) return "oops";
};
```

Or is this a question of implementation, e.g. passing *embedded-statement* to an extension method like `ForEach`?

**Comment by alrz** [comment]
Yeah, why not? 
```cs
do {
    // continue;
    // break;
    // return;
} while( ... );
```
This doesn't look magical:
```cs
from item in Enumerable.Range(0, 10) do {
    // continue;
    // break;
    // return;
};
```
or even
```cs
Enumerable.Range(0, 10) iterate item do { ... };
```

PS: Are we digging  `foreach` statement's grave?

**Comment by gafter** [comment]
@HaloFour There is no efficient way to pass am *embedded statement* to an extension method.

Its not that I'm against that sort of thing... see my talks about ["closures" for Java circa 2006-2007](https://www.youtube.com/watch?v=0zVizaCOhME&feature=youtu.be&t=1924) where I showed exactly how to do it and provided a language extension and implementation. I just don't know how to do it without a thousand-fold performance penalty on the CLR.

**Comment by HaloFour** [comment]
@alrz I wouldn't recommend going the route of trying to use an extension method anyway.  I was just fishing for what @gafter was intending by his question regarding not being able to break/return/etc from the loop.  I agree that if such a clause were to exist that it should just translate into a `foreach` loop:

`from x in ... foreach embedded-statement`

into:

`foreach (var x in ...) embedded-statement`

**Comment by gafter** [comment]
@HaloFour I see now what you had in mind. That does make sense to me.

**Comment by HaloFour** [comment]
@gafter Works for me.  I agree that the rest of the LINQ query still using the LINQ extension methods:

```cs
from item in new[] { "One", "Two", "Three" }
where item ==  "Two"
do Debug.WriteLine(item.ToString());

// equivalent to
var $temp = (new [] { "One", "Two", "Three" }).Where(item => item == "Two");
foreach (var item in $temp) Debug.WriteLine(item.ToString());
```

**Comment by aluanhaddad** [comment]
I like this idea. Placing the do clause at the end is nicely readable and keeps the scoping straightforward. 

One thing that occurs to me is that this could be quite useful as a more general kind of side effecting interjective operator, in the vein of Observable.Do.
However, that would seem to require that the do clause be implemented as a standard sequence operator, ruling out the use of break, continue, return, and so on, but personally I think that would be worth the trade off.

**Comment by HaloFour** [comment]
@aluanhaddad That would defeat the point of this proposal which is to avoid having to `foreach` over the resultant `IEnumerable<T>` separately.  You can already accomplish said operator with `select` assuming the expression would result in the same value as each element.

**Comment by aluanhaddad** [comment]
@HaloFour I see what you mean about the select operator, provided that you have a helper method which returns the value. But do could be implemented with semantics such that if the result is not assigned to a variable, and do is the last clause, then the Enumerable would be implicitly materialized. However as I think about it, that might be confusing.

Another nice thing about potentially making do a LINQ operator is that it could be translated into ParallelEnumerable.ForAll, something that foreach does not support. I think this would be a nice abstraction.  

That said, this is a simple and clean proposal and I do not wish to over complicate it

**Comment by TonyValenti** [comment]
Hi All,
I really like this idea and I love the syntax.  I would like to throw something at you to consider:

How would this work with PLINQ?  Does the ForEach become a regular "ForEach" or does it become a Parallel ForEach?

**Comment by aluanhaddad** [comment]
@TonyValenti I am also wondering about that possibility, see my comment above. 

One question though, I think you mean ParallelEnumerable.ForAll not Parallel.ForEach, as the latter is not directly related to PLINQ. Is that correct?

**Comment by gafter** [comment]
@aluanhaddad Thanks!

@TonyValenti If you use PLINQ, you'd get most of the query executed in parallel, but the last step (do) would end up being sequential.

**Comment by alrz** [comment]
@gafter "last step (do) would end up being sequential" I think that isn't  the case with async sequences.
```cs
await from item in asyncSeq async do await FooAsync(item);
```

**Comment by paulomorgado** [comment]
When LINQ was designed, it was designed as a pure query language without side effects. That's why [`Enumerable`](https://msdn.microsoft.com/library/system.linq.enumerable "Enumerable Class") doesn't have a `Do` or `ForEach` method and any mention of having such methods was frowned up on.

Is this:

    from item in new[] {"One", "Two", "Three" } 
            where item == "Two"
            do
    {
        Undo(item);
    }

So much better than:

    foreach(var item in from item in new[] {"One", "Two", "Three" }
                        where item == "Two"
                        select item)
    {
        Undo(item);
    }

that is worth all the trouble?

@dsaf's original proposal is very simplistic with it's direct projection of the enumerable item. How would that go with a projected transformation? Say, `select new Something(item)`. Would it be a `select ... do`?

**Comment by alrz** [comment]
@paulomorgado But a `foreach` statement will be needed per each `from`, see my example [above](https://github.com/dotnet/roslyn/issues/1938#issuecomment-156657721). In mix with LINQ operators, it totally worth it.

**Comment by paulomorgado** [comment]
@alrz, "totally worth it" means you'd be willing to drop something else. What? Would you favor this one instead of, say, #3571?

**Comment by alrz** [comment]
@paulomorgado While it has nothing to do with this proposal, the syntax you've proposed there, doesn't even make sense to me. `Distinct c.CountryID` this looks like F# function application btw. There is a similar conversation related to your proposal over at #100.

**Comment by paulomorgado** [comment]
So, you're operating in the assumption of infinite resources to implement zero-cost features. is it?

**Comment by HaloFour** [comment]
@paulomorgado I'd say it's up to the language design team to both determine whether or not a change like this fits within the philosophy of what LINQ is supposed to be (and whether or not that still applies) as to the priority of each of these proposals.  I don't think that anyone assumes that every proposal will be immediately considered and implemented.

As for `do` over an item projected by `select`, you'd need to use `select into` in order to project it into a separate range variable:

```cs
from item in new[] {"One", "Two", "Three" } 
        where item == "Two"
        select new Something(item) into item
        do
{
    Undo(item);
}
```

@aluanhaddad @TonyValenti As of now this `do embedded-statement` is translated directly into a `foreach` loop around the query.  So it would be sequential but also allow for control statements like `return`, `break` or `continue` which would not be possible if it translated into an extension method which accepted an `Action<T>`.

**Comment by aluanhaddad** [comment]
@HaloFour yes, it is a tradeoff. Personally, I would prefer the benefits of an extension method implementation, so I made some arguments for that, but it would still be useful if implemented as a direct translation to ```foreach```, so I'm in favor of it either way.

**Comment by weitzhandler** [comment]
https://github.com/dotnet/roslyn/issues/100#issuecomment-164934407

I strongly vote for this one.

I never use LINQ language queries and instead my coding guidelines is avoid using them and always use the extension methods directly, and that's because the ugly parentheses each query has to be wrapped in order to materialize it (`ToArray`, `ToList`, `Sum`, `SingleOrDefault` etc.).

Until this issue is addressed, language-build-in LINQ is merely useless.
I really hope to see this implemented soon. Maybe to a more limited extent (avoiding the introduction of a gazillion new language keywords.

I'd say the syntax should provide an operator that expects a shortcutted extension method available for `IEnumerable<TElement>`, for instance:

    //parents is Parent[]
    var parents = from student in students
                         where student.Age < 18
                         select student.Parent
                         call ToArray()

    //student is Student
    var student = from st in students
                          call SingleOrDefault(st => st.Id == id);

Asynchronous methods should also be supported:

       var student = from st in students
                             call await SingleOrDefaultAsync(st => st.Id == id);

Maybe there should be a verbose LINQ fashioned way to pass the arguments and completely avoid the use of parenthesis, but I personally don't see it as necessary.

Anyway this feature is crucial for the completeness of the LINQ syntax.
                            
Some suggestions above proposed the `ToList` at the beginning of the query, but that's a no go, since we want to be able to process it after the selection, and we don't want to be limited to parameterless ex. methods only. What if we wanna call `ToLookup` with a key selector, bottom line we can't tie it up to the language, we just need to find a way to call any `Enumerable` ex. methods on the current query state and treat its result as the final type of the query.

**Comment by TonyValenti** [comment]
Here is an idea:
What if we got a new operator?  I propose |. (Pipe dot).

var items = from x in source select x |.ToList();


On Tuesday, December 15, 2015, Shimmy <notifications@github.com> wrote:

> #100 (comment)
> <https://github.com/dotnet/roslyn/issues/100#issuecomment-164934407>
>
> I strongly vote for this one.
>
> I never use LINQ language queries and instead my coding guidelines is
> avoid using them and always use the extension methods directly, and that's
> because the ugly parentheses each query has to be wrapped in order to
> materialize it (ToArray, ToList, Sum, SingleOrDefault etc.).
>
> Until this issue is addressed, language-build-in LINQ is merely useless.
> I really hope to see this implemented soon. Maybe to a more limited extent
> (avoiding the introduction of a gazillion new language keywords.
>
> I'd say the syntax should provide an operator that expects a shortcutted
> extension method available for IEnumerable<TElement>, for instance:
>
> //parents is Parent[]
> var parents = from student in students
>                      where student.Age < 18
>                      select student.Parent
>                      call ToArray()
>
> //student is Student
> var student = from st in students
>                       call SingleOrDefault(st => st.Id == id);
>
> Asynchronous methods should also be supported:
>
>    var student = from st in students
>                          call await SingleOrDefaultAsync(st => st.Id == id);
>
> Maybe there should be a verbose LINQ fashioned way to pass the arguments
> and completely avoid the use of parenthesis, but I personally don't see it
> as necessary.
>
> Anyway this feature is crucial for the completeness of the LINQ syntax.
>
> Some suggestions above proposed the ToList at the beginning of the query,
> but that's a no go, since we want to be able to process it after the
> selection, and we don't want to be limited to parameterless ex. methods
> only. What if we wanna call ToLookup with a key selector, bottom line we
> can't tie it up to the language, we just need to find a way to call any
> Enumerable ex. methods on the current query state and treat its result as
> the final type of the query.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dotnet/roslyn/issues/1938#issuecomment-164987826>.
>


-- 
Tony Valenti

**Comment by alrz** [comment]
@TonyValenti #100.

**Comment by dsaf** [comment]
@TonyValenti @alrz actually sounds like forward pipe with a special case for LINQ to prevent parentheses #5445. Not sure an operator would be idiomatic, since they have already introduced ```equals``` in addition to ```==```.

**Comment by alrz** [comment]
@dsaf on that idea, see [#100 (comment)](https://github.com/dotnet/roslyn/issues/100#issuecomment-157729580).

**Comment by alrz** [comment]
It'd be nice to differ the semicolon requirement for the *embedded-statement* in `do` clause like
```
F(list => from foo in list do WriteLine(foo.Bar));
```
because *query-expression* is an expression afterall.

/cc @HaloFour

**Comment by gafter** [comment]
@alrz I think this proposal is for a new statement form, not a new expression form.

My working vision of this is in https://github.com/dotnet/roslyn/issues/1938#issuecomment-156730804

**Comment by alrz** [comment]
@gafter So in that case, can we use `let` with patterns in this new LINQ statement form, right? Because then it just would be translate to a `let` statement (above example).

**Comment by alrz** [comment]
@gafter So in that case, `let` can be simply translated to a `let` statement, unlike #6877.

**Comment by MovGP0** [comment]
instead of 
```csharp
foreach (var item in from xItem in new[] {"One", "Two", "Three" } 
                     where xItem == "Two"
                     select xItem)
{
    Debug.WriteLine(item);
}
```
you should use [MoreLINQ](https://github.com/morelinq) and [Command Query Separation (CQS)](https://martinfowler.com/bliki/CommandQuerySeparation.html) and write it like this: 
```csharp
// Query
var items = new[] {"One", "Two", "Three" }.Select(i => i == "Two");

// Command 
items.ForEach(Debug.WriteLine);
```

**Comment by aluanhaddad** [comment]
@MovGP0 you mean to get

```C#
public static void ForEach<T>(this IEnumerable<T> source, Action<T> action)
{
    foreach(var element in source)
    {
        action(source);
    }
}
```
I've tried it, added it to many projects, but I do not like how it feels as compared to `foreach`. Basically, I always end up wanting to write a chainable version of it, like `Observable.Do`, and here be dragons.

Also, that assumes you only need to support method syntax. I prefer query syntax where available, but I like both.

**Comment by MovGP0** [comment]
@aluanhaddad doesn't really matter which syntax you use, as long as you stick to CQS. Using another syntax only means that you have more lines of code to type: 
```csharp
// query
var items = from item in new[] {"One", "Two", "Three" } 
            where item == "Two"
            select item;

// command
foreach (var item in items)
{
    Debug.WriteLine(item);
}

**Comment by bondsbw** [comment]
@MovGP0 That depends on the syntax; several examples here would have reduced the number of lines.

**Comment by gafter** [comment]
@MovGP0 How do I `break`, `continue`, etc?

**Comment by MovGP0** [comment]
@gafter by putting the query into a separate (extension) method, which returns an `IEnumerable`, and using `yield`/`break`/`continue` statements in an `foreach` loop there. 

Do you have a code example?

**Comment by TonyValenti** [comment]
Related to "How do I break or continue?" I think it is worth deciding what the return value from "do" is.  Is a linq statement with "do" a statement or an expression?

My preference is that "do" act like either an Action<> or a Func<> and return either:
IEnumerable<T> when "yield returns/Yield Breaks" are used in the DO.
T or Void when "Return or Breaks".

Here are a few different ways the syntax could look for different things:
````
//This returns a List<String>
var FirstNotTwo =
from item in new[] {"One", "Two", "Three" } 
where item != "Two"
do ToList();

//This emulates a FOR loop and returns Void:
from item in new[] {"One", "Two", "Three" } 
do (var Index) {
  Console.WriteLine($"Item {Index}: {item}");
}


//This returns void:
from item in new[] {"One", "Two", "Three" } 
where item == "Two"
do {
    Debug.WriteLine(item.ToString());
}

//This returns an IEnumerable<String>
var NotTwo =
from item in new[] {"One", "Two", "Three" } 
do {
    if(item == "Two") {
        yield break;
    } else {
        yield return item;
    }
}

//This returns an IEnumerable<String>
var NotTwo =
from item in new[] {"One", "Two", "Three" } 
do {
    if(item == "Two") {
        yield break;
    } else {
        yield return item;
    }
}

//This returns a String
var FirstNotTwo =
from item in new[] {"One", "Two", "Three" } 
do {
    if(item != "Two") {
        return item;
    }
}


````

**Comment by MovGP0** [comment]
@TonyValenti  That `from ... do` syntax merges the Query with the Command (and thus violates the CQS principle), and in many cases is more to write. I find it cleaner this way:  

```csharp
var ott = new[] {"One", "Two", "Three" };
```

```csharp
//This returns a List<String>
var firstNotTwo = ott.Where(item => item != "Two").ToList();
```

```csharp
//This emulates a FOR loop and returns void:
// query
var indexedItems = ott.Select((item, index) => new { Item = item, Index = index });

// command
indexedItems.ForEach(i => Console.WriteLine($"Item {i.Index}: {i.Item}");
```

```csharp
//This returns void:
// query
var two = ott.Single(item => item == "Two");

// command
Debug.WriteLine(two);
```

```csharp
//This returns an IEnumerable<String>
var notTwo = ott.Where(item => item != "Two");
```

```csharp
//This returns an IEnumerable<String>
var notTwo = ott.TakeWhile(item => item != "Two");
```

```csharp
//This returns a String
var firstNotTwo = ott.First(item => item != "Two");
```

**Comment by bondsbw** [comment]
@MovGP0 There may always be a great debate about whether query syntax is better than method chaining, but this proposal is to make query syntax better for those who prefer it. (Though I would love to see a `ForEach(...)` method ship in the BCL.)

Besides, I don't think executing two statements directly in succession (in the same block of code) is really any more/less compliant with the spirit of CQS than a single statement would be.  Just my 2 cents.

**Comment by jnm2** [comment]
@bondsbw You're completely right. I wasn't going to say anything but the CQS thing is being taken way out of context in my opinion.

**Comment by gafter** [comment]
> I think it is worth deciding what the return value from "do" is.

The proposed construct is intended to be a statement form.

**Comment by gafter** [comment]
Now tracked at https://github.com/dotnet/csharplang/issues/101
